compiler/rustc_mir_build/src/build/scope.rs

```diff
@@ -783,6 +788,46 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         }
     }
 
+    /// Walks upwards from `orig_id` to find a node which might change lint levels with attributes.
+    /// It stops at `self.hir_id` and just returns it if reached.
+    fn maybe_lint_level_root_bounded(&mut self, orig_id: HirId) -> HirId {
+        let mut id = orig_id;
+        let hir = self.tcx.hir();
+        loop {
+            if id == self.hir_id {
+                // This is a moderately common case, mostly hit for previously unseen nodes.
+                break;
+            }
+
+            if hir.attrs(id).iter().any(|attr| Level::from_attr(attr).is_some()) {
+                // This is a rare case. It's for a node path that doesn't reach the root due to an
+                // intervening lint level attribute. This result doesn't get cached.
+                return id;
```

Comment:
Should we eventually cache this too? If we have a whole HIR subtree that hits the same lint root, different than `self.hir_id`?

Reply:
I originally did cache these values, using an `FxHashMap<HirId, HirId>`. Then I realized that 99% of the values stored were `self.hir_id`, which seemed wasteful. So I tried changing it to `FxHashSet<HirId>` and only caching those 99% and the instruction count dropped very slightly, while also using less memory. And if we want to use  bitset (like you suggested above) then we'll need to keep this design.